### PR TITLE
Allow to use some system libraries instead of bundled

### LINKED ISSUE #16948: Use System FLAC
### Your idea

Add the ability to use the native FLAC available on the system

### Problem to be solved

So far it is only possible to build Musescore with the bundled FLAC library.

### Prior art

_No response_

### Additional context

This is my first attempt to improve Musescore build system. I'm a beginner so be indulgent. I am listening to all your comments. This patch has been created for Musescore v4.0.2. 

```
From 9a49c77a6edbcd8d37e1890e8b5af7dd67c06cee Mon Sep 17 00:00:00 2001
From: "Jan-Michał T.Dydak" <jm.dev@gmx.com>
Date: Mon, 20 Mar 2023 18:47:10 +0100
Subject: [PATCH] use-system-flac

Add the ability to use the native FLAC available on the system
---
 CMakeLists.txt                     |  1 +
 build/cmake/FindFLAC.cmake         | 40 ++++++++++++++++++++++++++++++
 src/framework/audio/CMakeLists.txt | 11 ++++++--
 3 files changed, 50 insertions(+), 2 deletions(-)
 create mode 100644 build/cmake/FindFLAC.cmake

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 43139ec9f4..2a1c18718d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -106,6 +106,7 @@ endif()
 
 set(QT_SUPPORT ON)
 
+option(USE_SYSTEM_FLAC "Use system FLAC" OFF)
 option(USE_SYSTEM_FREETYPE "Use system FreeType" OFF) # requires freetype >= 2.5.2, does not work on win
 set(SCRIPT_INTERFACE  TRUE)
 
diff --git a/build/cmake/FindFLAC.cmake b/build/cmake/FindFLAC.cmake
new file mode 100644
index 0000000000..2a2a2bd0d1
--- /dev/null
+++ b/build/cmake/FindFLAC.cmake
@@ -0,0 +1,40 @@
+## Find FLAC
+## Find the native FLAC includes and libraries
+## --------------------------------------------
+
+find_package (PkgConfig QUIET)
+pkg_check_modules (PCFLAC QUIET flac>=1.3.4)
+
+set (FLAC_VERSION ${PCFLAC_VERSION})
+
+find_path (FLACPP_INCLUDE_DIR
+   NAMES "FLAC++/encoder.h"
+   DOC   "Header path required to include <FLAC++/encoder.h>")
+
+find_library (FLACPP_LIBRARY
+   NAMES "FLAC++"
+   DOC   "Location of FLAC++ library")
+
+find_library (FLAC_LIBRARY
+   NAMES "FLAC"
+   DOC   "Location of FLAC library")
+
+include (FindPackageHandleStandardArgs)
+
+find_package_handle_standard_args (FLAC
+   REQUIRED_VARS
+         FLACPP_LIBRARY
+         FLACPP_INCLUDE_DIR
+   VERSION_VAR
+         FLAC_VERSION)
+
+find_package_handle_standard_args (FLAC
+   REQUIRED_VARS
+         FLAC_LIBRARY
+   VERSION_VAR
+         FLAC_VERSION)
+
+if (FLAC_FOUND)
+   set (FLAC_INCLUDE_DIRS ${FLACPP_INCLUDE_DIR})
+   set (FLAC_LIBRARIES ${FLACPP_LIBRARY} ${FLAC_LIBRARY})
+endif ()
diff --git a/src/framework/audio/CMakeLists.txt b/src/framework/audio/CMakeLists.txt
index 7cde6816bd..ac95f5eb97 100644
--- a/src/framework/audio/CMakeLists.txt
+++ b/src/framework/audio/CMakeLists.txt
@@ -218,9 +218,16 @@ if (ENABLE_AUDIO_EXPORT)
 
     add_subdirectory(${PROJECT_SOURCE_DIR}/thirdparty/lame lame)
     add_subdirectory(${PROJECT_SOURCE_DIR}/thirdparty/opusenc opusenc)
-    add_subdirectory(${PROJECT_SOURCE_DIR}/thirdparty/flac flac)
 
-    set(MODULE_LINK ${MODULE_LINK} lame opusenc flac)
+    if (USE_SYSTEM_FLAC)
+        find_package(FLAC REQUIRED)
+        set(MODULE_LINK ${MODULE_LINK} ${FLAC_LIBRARIES})
+    else()
+        add_subdirectory(${PROJECT_SOURCE_DIR}/thirdparty/flac flac)
+        set(MODULE_LINK ${MODULE_LINK} flac)
+    endif()
+
+    set(MODULE_LINK ${MODULE_LINK} lame opusenc)
 endif()
 
 if (OS_IS_MAC)
-- 
2.39.2
```

[0001-use-system-flac.tar.gz](https://github.com/musescore/MuseScore/files/11021812/0001-use-system-flac.tar.gz)

### PR DESCRIPTION
**Pull request by doronbehar**
Help Linux package maintainers with using system libraries instead of bundled `thirdparty/` libraries. See discussion at #11572. This also resolves (almost) #16948.

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, **Haven't tested each commit individually, but should compile as well**.
- [ ] I created a unit test or vtest to verify the changes I made (if applicable)

### CODE REVIEW

File: src/framework/audio/internal/encoders/oggencoder.cpp
Comment (bluebear94) [inline src/framework/audio/internal/encoders/oggencoder.cpp:26]: Wouldn’t it be less confusing to use `<>` for including system header files?

```suggestion
#include <opus/opusenc.h>
```

**Comment by cbjeukendrup** [comment]
I had an idea: can we put all the logic for finding these libraries or building them from source in custom `Find<PackageName>.cmake` modules (see https://cmake.org/cmake/help/latest/manual/cmake-developer.7.html#find-modules)? 

In that module, we would write the following logic:
- if MUE_COMPILE_USE_SYSTEM_… is true:
   - try to find the package via pkgconfig
   - If successful: 
      - that gives a PkgConfig::… imported target
      - Create an alias target called something like `ThirdParty::…` to that imported target. 
      - call `return()`.
- If the code after that gets reached, that's either because `MUE_COMPILE_USE_SYSTEM_…` was false or because it was not found by PkgConfig. 
- So, we'll create a target that builds the library from the sources included in the repo. 
- Create an alias target also called `ThirdParty::…` to that created target

Now, using the library is as simple as doing `find_package(…)`, followed by `target_link_libraries(…)`, or the latter rather via `set(MODULE_LINK …)`.
(target_link_libraries does apparently not only cause the libraries to be linked, but also sets include directories etc transitively.)

I'm not 100% sure that it will work, and it might be slightly more work, but if this approach succeeds it would be quite neat.

File: src/framework/audio/internal/encoders/oggencoder.cpp
Comment (doronbehar) [reply src/framework/audio/internal/encoders/oggencoder.cpp:26]: > Wouldn’t it be less confusing to use `<>` for including system header files?

This worked, also allowing the `set(MODULE_INCLUDE` to be removed. Thanks!

**Comment by doronbehar** [comment]
> If successful:
>   * that gives a PkgConfig::… imported target
>   * Create an alias target called something like `ThirdParty::…` to that imported target.
>   * call `return()`.

I'm sorry @cbjeukendrup but I am not experienced enough with cmake scripting to produce this code...

I pushed @bluebear94's suggestion, and I'm debugging the CI failure now.

**Comment by doronbehar** [comment]
CI is hopefully fixed now.

**Comment by doronbehar** [comment]
Additionally, I almost managed to make musescore build with a system installed kddockwidgets, but here on NixOS and at upstream, kddockwidgets' latest version is 1.7.0 and here it's 1.4.95. Many header files have changed their paths.. I managed to guess some of the renames, but I wasn't sure about the rest.

Hence, I won't push that to this PR, but I pushed that attempt to [this branch](https://github.com/musescore/MuseScore/compare/master...doronbehar:MuseScore:system-kddockwidgets?expand=1). I also updated my TODO list at https://github.com/musescore/MuseScore/issues/11572#issuecomment-1773742874 . As I said in https://github.com/musescore/MuseScore/issues/18795#issuecomment-1671609633 , I think it'd be best for you maintainers to support only the latest version of such a dependency (update both the bundled version and the required one in `find_package`), and not use complicated `#if` - `#else` processors to accommodate with those incompatible versions.

**Comment by doronbehar** [comment]
Friendly ping @cbjeukendrup

**Comment by cbjeukendrup** [comment]
I'll look into this next week, when I have Christmas holiday :) 
(I would namely still like to try if it's possible to implement this according to my suggestion above)

**Comment by doronbehar** [comment]
Oh I forgot about your suggestion from back then my cmake skills haven't improved since :). I was only wondering what is with this PR, due to the last release, hopefully it is still worth reconsidering in it's current form. Marry Christmas :christmas_tree: .

**Comment by doronbehar** [comment]
Hello @cbjeukendrup . I pushed a rebase of this PR, along with an additional PR that puts the logic of finding tinyxml in a dedicated `FindTinyXml.cmake` file. I can write 2 more such commits for `FindFlac.cmake` and `FindLibopusenc.cmake`, but it'd be nice to get your approval on the implementation. I'm not sure it is good because it doesn't feel as if I simplified the `src/framework/global/CMakeLists.txt` file.

I'm testing this on my machine now, with Nixpkgs' build system, with and without the external tinyxml, and I'm working on fixing CI as well.

**Review by igorkorsukov — APPROVED** [review]

**Comment by doronbehar** [comment]
Thank you @igorkorsukov !

**Comment by doronbehar** [comment]
This wasn't picked up into 4.3 :(. Is it planned to be picked up in a patch version? Or 4.4?

**Comment by Jojo-Schmitz** [comment]
More likely 4.4

**Comment by cbjeukendrup** [comment]
@doronbehar I discussed it within the team, and since this change wouldn't have any effect on the builds that _we_ create, we decided that if it's important for anyone, it's okay to retroactively add this change to the 4.3 branch and move the release tag. So if it's indeed important, could you open a PR for the 4.3 branch?

**Comment by Jojo-Schmitz** [comment]
Moving that tag is needed anyhow, as currently is is one commit behind already

**Comment by doronbehar** [comment]
It's not that urgent for us :) I was only expecting this change to land in the 4.3.0 release, and I wanted to know to when should I schedule the update to Nixpkgs' Nix expression. If you won't mind backporting this to the 4.3 branch, sure, it'd be great :). PR is in https://github.com/musescore/MuseScore/pull/22724 .

**Pull request merged**
